Fix: webp files corrupted by ExifInterface.saveAttributes

CoXier opened the pull request:
## Proposed Changes

  - webp file without vp8x header (maybe old format ? )
  - width or height exceeds 8191
  - ExifInterface.saveAttributes

## Testing

Test: Describe how you tested your changes. Note that this line (with `Test:`) is required, your PR will not build without it!

## Issues Fixed

Fixes: The bug on [34269705](https://issuetracker.google.com/issues/342697059) being fixed

CoXier committed:
Fix: webp files corrupted by ExifInterface.saveAttributes when width or height exceeds 8191.

More details: https://issuetracker.google.com/issues/342697059

CoXier commented:
Hi @dlam @yigit 

Could you please provide an update on its review status? Thank you!

icbaker commented on exifinterface/exifinterface/src/main/java/androidx/exifinterface/media/ExifInterface.java:
```diff
@@ -6728,8 +6730,11 @@ private void saveWebpAttributes(InputStream inputStream, OutputStream outputStre
+                        int width = widthAndHeight & 0x0000FFFF;
+                        int height = (widthAndHeight & 0xFFFF0000) >> 16;
+                        if (width > WEBP_MAX_DIMENSION || height > WEBP_MAX_DIMENSION) {
+                            throw new IOException("Maximum side allowed is 16383 pixels.");
```
Reading https://datatracker.ietf.org/doc/html/rfc6386#section-9.1 it looks like this is really asserting that both the horizontal and vertical scaling factors are zero (meaning no upscaling).

The current code looks like it aims to ignore the scaling factor instead, and this seems to match this wording in the WebP spec:
> Note that the VP8 frame header contains the VP8 frame width and height. That is assumed to be the width and height of the canvas.

Assuming that "frame width" refers to the 14 bits of width in the VP8 frame, ignoring the scaling factor (which seems to fit the wording of rfc6386 too.

I'd suggest changing your code to ignore the scaling factor too - I think you can most easily do that by amending your bit masks to only read 14 bits each (and then no need to check the size, since you can't read more than 16383 if you only read 14 bits).

CoXier replied:
I think we should inform the caller about the maximum supported dimensions. If we only read 14 bits, the caller won't be able to perceive any abnormality.

icbaker replied:
Not sure I understand. Each dimension is only permitted to be 14 bits, the other 2 bits of the 16 are not dimensions, they are a scaling factor that isn't relevant here, so not reading these 2 bits seems correct to me.

CoXier replied:
Thank you for your review. After carefully studying the RFC 6386 document and the VP8 specification, I realize that my initial understanding was incorrect. You are right; the width and height should indeed only use 14 bits, not 16.

I agree with your suggestion to modify the code to read only 14 bits for width and height. This not only more accurately follows the specification but also automatically limits the maximum dimensions to 16383x16383 pixels without requiring additional checks.

I will update the code according to your recommendation, using 14-bit masks to read the width and height. Additionally, I think it would be beneficial to explicitly state this maximum size limitation in the documentation to ensure callers are aware of this constraint.

icbaker replied:
> Additionally, I think it would be beneficial to explicitly state this maximum size limitation in the documentation to ensure callers are aware of this constraint.

I don't think it's necessary for `ExifInterface`'s documentation to duplicate information that is available in each image format spec (as is the case here). Otherwise it risks getting very unwieldy - since we end up trying to include all the information from the specs of the many image formats we support.

CoXier replied:
I updated my code just now. Please have a look.

icbaker commented:
Please can you add a test image and test case that fails without this fix?

Maybe you could make a copy of [`webp_without_exif.webp`](https://github.com/androidx/androidx/blob/androidx-main/exifinterface/exifinterface/src/androidTest/res/raw/webp_without_exif.webp) and resize it to 8192 or greater in one dimension (just be careful that whatever tool you use to resize doesn't "usefully" add a VP8X or EXIF chunk).

CoXier commented:
> Please can you add a test image and test case that fails without this fix?
> 
> Maybe you could make a copy of [`webp_without_exif.webp`](https://github.com/androidx/androidx/blob/androidx-main/exifinterface/exifinterface/src/androidTest/res/raw/webp_without_exif.webp) and resize it to 8192 or greater in one dimension (just be careful that whatever tool you use to resize doesn't "usefully" add a VP8X or EXIF chunk).

I think this bug is very clear (I explain why this happens). I'm not familiar how to add test case to this repo and I can provide a test image.

icbaker commented:
I've added a test that exercises this bug, and marked it `@Ignore`, and sent it for review: http://r.android.com/3204431

I think there are a few options:
1. You copy what I've done there into this PR (I included the command I used to generate the image) and I drop that change
2. You wait for that to be merged, then rebase on top of it and remove the `@Ignore` in this PR
3. You drop this PR and I merge the fix into the same change  as the test (this is probably the quickest option)

CoXier commented:
Thanks  for  your  work. 2nd option  is  better for  me.

In  the past few months , I tried to  build  the  androidx repo using
Android studio , however I failed. Because of that，I cannot write the test
for this fix.

Thanks a lot  for  your  work，this is my first  pr for  Google  Android.

Ian Baker ***@***.***>于2024年8月1日 周四23:23写道：

> I've added a test that exercises this bug, and marked it @Ignore, and
> sent it for review: http://r.android.com/3204431
>
> I think there are a few options:
>
>    1. You copy what I've done there into this PR (I included the command
>    I used to generate the image) and I drop that change
>    2. You wait for that to be merged, then rebase on top of it and remove
>    the @Ignore in this PR
>    3. You drop this PR and I merge the fix into the same change as the
>    test (this is probably the quickest option)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/androidx/androidx/pull/683#issuecomment-2263345045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEPBALJZHUUMSKNL2UB2QATZPJHF3AVCNFSM6AAAAABIJMTDAWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENRTGM2DKMBUGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

CoXier commented:
> I've added a test that exercises this bug, and marked it `@Ignore`, and sent it for review: http://r.android.com/3204431
> 
> I think there are a few options:
> 
> 1. You copy what I've done there into this PR (I included the command I used to generate the image) and I drop that change
> 2. You wait for that to be merged, then rebase on top of it and remove the `@Ignore` in this PR
> 3. You drop this PR and I merge the fix into the same change  as the test (this is probably the quickest option)

Thank you for your work. The second option works better for me.

Over the past few months, I attempted to build the AndroidX repository using Android Studio, but unfortunately, I was unsuccessful. As a result, I am unable to write tests for this fix.

I really appreciate your efforts; this is my first pull request for Google Android.

CoXier committed:
Fix: enable test for https://issuetracker.google.com/342697059

CoXier commented:
@icbaker I already rebase the androidx-main branch and enable the test.

icbaker commented:
Thanks, please can you take a look at my comment on L6736 of your code as well.

CoXier commented:
> Thanks, please can you take a look at my comment on L6736 of your code as well.

Where can I find the comment ? Github ?

icbaker commented:
> > Thanks, please can you take a look at my comment on L6736 of your code as well.
> 
> Where can I find the comment ? Github ?

It helps if i actually send it - sorry...

CoXier committed:
Opt: Refine WebP width and height extraction to 14 bits

- Modify bit masks to extract only 14 bits for both width and height
- Remove 16-bit masking, now using 0x3FFF (14 bits) for both dimensions
- Align implementation with WebP spec and RFC 6386

This change ensures we're reading the actual frame dimensions from the
VP8 frame header, ignoring the scaling factor. It matches the WebP spec
which states that the VP8 frame width and height are assumed to be the
width and height of the canvas.

Ref: https://datatracker.ietf.org/doc/html/rfc6386#section-9.1

icbaker commented:
Looks good to me, @dlam or @yigit please can you take a look and merge this?

icbaker commented:
> Looks good to me, @dlam or @yigit please can you take a look and merge this?

I misunderstood the process - it's being merged in http://r.android.com/3204842 which will then close this PR once it's submitted.

icbaker commented on exifinterface/exifinterface/src/main/java/androidx/exifinterface/media/ExifInterface.java:
```diff
@@ -6774,8 +6774,8 @@ private void saveWebpAttributes(InputStream inputStream, OutputStream outputStre
-                        width = (widthAndHeight << 18) >> 18;
-                        height = (widthAndHeight << 2) >> 18;
+                        width = widthAndHeight & 0x3FFF;
+                        height = (widthAndHeight >> 16) & 0x3FFF
```
Please can you add the missing semi-colon at the end of this line

CoXier replied:
Sorry for my mistake. I love Kotlin lol

icbaker commented:
I'll take a look at the test failure on http://r.android.com/3204842 tomorrow - it looks like my test causes an OOM when trying to decode the large WebP file to check it's still valid.

icbaker commented:
The test is only failing on API 21 on the [FTL](https://firebase.google.com/docs/test-lab) emulators used in the automated testing (I wasn't able to repro the same failre on an API 21 x86 emulator launched from Android Studio locally).

I suggest annotating this test with the following line:

```
@SdkSuppress(minSdkVersion = 22) // Parsing the large image causes OOM on API 21 FTL emulators.
```

CoXier commented:
> The test is only failing on API 21 on the [FTL](https://firebase.google.com/docs/test-lab) emulators used in the automated testing (I wasn't able to repro the same failre on an API 21 x86 emulator launched from Android Studio locally).
> 
> I suggest annotating this test with the following line:
> 
> ```
> @SdkSuppress(minSdkVersion = 22) // Parsing the large image causes OOM on API 21 FTL emulators.
> ```

I wonder whether the full classname is androidx.test.filters.SdkSuppress?

CoXier committed:
Test: Add SDK version suppression for WebP parsing test

- Add @SdkSuppress(minSdkVersion = 22) annotation to the test
- Exclude test execution on API levels below 22

This change addresses an out-of-memory (OOM) issue occurring
specifically on API 21 Firebase Test Lab (FTL) emulators when
parsing large WebP images. The test continues to run on API 22
and above, ensuring compatibility with the majority of target
devices while avoiding false negatives on older API levels.

Note: The OOM issue was not reproducible on local API 21 x86
emulators, suggesting it may be specific to FTL environments.

CoXier commented:
@icbaker Now that my code changes have been merged, I was wondering if there's an estimated timeline for the next release of androidx.exifinterface:exifinterface that would include these updates.

icbaker commented:
We are planning a 1.4.0 release, but I'm afraid we don't have a timeline to share at the moment.
